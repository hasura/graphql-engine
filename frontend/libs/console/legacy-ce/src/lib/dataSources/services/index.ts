import { postgres } from './postgresql';
import { mysql } from './mysql';
import { mssql } from './mssql';
import { bigquery } from './bigquery';
import { citus } from './citus';
import { cockroach } from './cockroach';

export const services = {
  postgres,
  mysql,
  mssql,
  bigquery,
  citus,
  cockroach,
};
