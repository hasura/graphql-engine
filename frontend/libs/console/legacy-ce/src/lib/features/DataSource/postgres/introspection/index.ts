export { getTrackableTables } from './getTrackableTables';
export { getDatabaseConfiguration } from './getDatabaseConfiguration';
export { getTableColumns } from './getTableColumns';
export { getFKRelationships } from './getFKRelationships';
export { getTablesListAsTree } from './getTablesListAsTree';
export { getSupportedOperators } from './getSupportedOperators';
export { getTrackableFunctions } from './getTrackableFunctions';
export { getDatabaseSchemas } from './getDatabaseSchemas';
export { getIsTableView } from './getIsTableView';
