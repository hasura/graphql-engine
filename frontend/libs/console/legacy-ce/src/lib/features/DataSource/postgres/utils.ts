import { getEntries } from '../../../components/Services/Data/Common/tsUtils';
import { RunSQLResponse } from '../api';
import { IntrospectedTable, TableColumn } from '../types';

function containsUppercase(str: string) {
  return /[A-Z]/.test(str);
}

export function adaptStringForPostgres(str: string) {
  return containsUppercase(str) ? `"${str}"` : str;
}

export const consoleDataTypeToSQLTypeMap: Record<
  TableColumn['consoleDataType'],
  string[]
> = {
  boolean: ['boolean', 'bool'],
  string: [
    'box',
    'character',
    'character varying',
    'circle',
    'line',
    'lseg',
    'macaddr',
    'macaddr8',
    'path',
    'pg_lsn',
    'pg_snapshot',
    'point',
    'polygon',
    'tsquery',
    'tsvector',
    'txid_snapshot',
    'uuid',
    'char',
    'varchar',
    'date',
    'time',
    'timetz',
    'timestamptz',
    'timestamp',
  ],
  text: ['text'],
  number: [
    'bigint',
    'bigserial',
    'bit',
    'bit varying',
    'bytea',
    'cidr',
    'double precision',
    'inet',
    'integer',
    'money',
    'numeric',
    'real',
    'smallint',
    'smallserial',
    'serial',
    'int8',
    'serial8',
    'varbit',
    'int',
    'int4',
    'decimal',
    'float4',
    'int2',
    'serial2',
    'serial4',
  ],
  json: ['interval', 'json', 'jsonb', 'xml'],
  float: ['float8'],
};

export const consoleScalars = Object.values(consoleDataTypeToSQLTypeMap).flat();

export function adaptSQLDataType(
  sqlDataType: string
): TableColumn['consoleDataType'] {
  const [dataType] = getEntries(consoleDataTypeToSQLTypeMap).find(([, value]) =>
    value.includes(sqlDataType)
  ) ?? ['string', []];

  return dataType;
}

export const adaptIntrospectedTables = ([
  fullTableListResult,
  partitionsTablesResult,
]: RunSQLResponse[]): IntrospectedTable[] => {
  const partitionNames =
    partitionsTablesResult.result?.map(row => row[0]) ?? [];
  /* 
    The `slice(1)` on the result is done because the first item of the result is always the columns names from the SQL output.
    It is not required for the final result and should be avoided 
  */
  const adaptedResponse = fullTableListResult?.result
    ?.slice(1)
    .map((row: string[]) => ({
      name: `${row[1]}.${row[0]}`,
      table: {
        name: row[0],
        schema: row[1],
      },
      type: partitionNames.includes(row[0]) ? 'PARTITION' : row[2],
    }));

  return adaptedResponse ?? [];
};
