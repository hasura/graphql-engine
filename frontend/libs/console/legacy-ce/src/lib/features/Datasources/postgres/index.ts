import * as hooks from './hooks';

export type { PostgresDataTarget } from './types';

export default {
  ...hooks,
};
