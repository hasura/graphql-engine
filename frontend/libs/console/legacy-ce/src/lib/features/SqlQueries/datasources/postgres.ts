import type { TableORSchemaArg } from '@/dataSources/types';
import { QualifiedTable } from '@/metadata/types';
import type { DatasourceSqlQueries } from '.';
import { getSchemasWhereClause, getTablesWhereClause } from './common';

const getKeysSql = (
  type: 'PRIMARY KEY' | 'UNIQUE',
  options: TableORSchemaArg
) => `
  -- test_id = ${'schemas' in options ? 'multi' : 'single'}_${
  type === 'UNIQUE' ? 'unique' : 'primary'
}_key
  SELECT
  COALESCE(
    json_agg(
      row_to_json(info)
    ),
    '[]' :: JSON
  )
  FROM (
      SELECT
          tc.table_name,
          tc.constraint_schema AS table_schema,
          tc.constraint_name,
          json_agg(kcu.column_name) AS columns
      FROM
          information_schema.table_constraints tc
          JOIN information_schema.key_column_usage kcu USING (constraint_schema, constraint_name)
    ${
      'schemas' in options
        ? getSchemasWhereClause(options.schemas)('tc.constraint_schema')
        : getTablesWhereClause(options.tables)({
            name: 'tc.table_name',
            schema: 'tc.constraint_schema',
          })
    }
          AND tc.constraint_type::text = '${type}'::text
      GROUP BY
          tc.table_name,
          tc.constraint_schema,
          tc.constraint_name) AS info;
  `;

const getTableColumnsSql = ({ name, schema }: QualifiedTable) => {
  if (!name || !schema) throw Error('empty parameters are not allowed!');

  return `SELECT table_catalog as database, table_schema, table_name, column_name, data_type FROM information_schema.columns WHERE table_schema = '${schema}' AND table_name  = '${name}';`;
};

export const postgresSqlQueries: DatasourceSqlQueries = {
  getFetchTablesListQuery(options: TableORSchemaArg): string {
    return `
  -- test_id = ${'schemas' in options ? 'multi' : 'single'}_table
  SELECT
    COALESCE(Json_agg(Row_to_json(info)), '[]' :: json) AS tables
  FROM (
    WITH partitions AS (
      SELECT array(
        WITH partitioned_tables AS (SELECT array(SELECT oid FROM pg_class WHERE relkind = 'p') AS parent_tables)
        SELECT
        child.relname       AS partition
    FROM partitioned_tables, pg_inherits
        JOIN pg_class child             ON pg_inherits.inhrelid   = child.oid
        JOIN pg_namespace nmsp_child    ON nmsp_child.oid   = child.relnamespace
    ${
      'schemas' in options
        ? getSchemasWhereClause(options.schemas)('nmsp_child.nspname')
        : getTablesWhereClause(options.tables)({
            name: 'child.relname',
            schema: 'nmsp_child.nspname',
          })
    }
    AND pg_inherits.inhparent = ANY (partitioned_tables.parent_tables)
      ) AS names
    )
    SELECT
      pgn.nspname AS table_schema,
      pgc.relname AS table_name,
      CASE
        WHEN pgc.relkind = 'r' THEN 'TABLE'
        WHEN pgc.relkind = 'f' THEN 'FOREIGN TABLE'
        WHEN pgc.relkind = 'v' THEN 'VIEW'
        WHEN pgc.relkind = 'm' THEN 'MATERIALIZED VIEW'
        WHEN pgc.relkind = 'p' THEN 'PARTITIONED TABLE'
      END AS table_type,
      obj_description(pgc.oid) AS comment,
      COALESCE(json_agg(DISTINCT row_to_json(isc) :: jsonb || jsonb_build_object('comment', col_description(pga.attrelid, pga.attnum))) filter (WHERE isc.column_name IS NOT NULL), '[]' :: json) AS columns,
      COALESCE(json_agg(DISTINCT row_to_json(ist) :: jsonb || jsonb_build_object('comment', obj_description(pgt.oid))) filter (WHERE ist.trigger_name IS NOT NULL), '[]' :: json) AS triggers,
      row_to_json(isv) AS view_info
      FROM partitions, pg_class as pgc  
      INNER JOIN pg_namespace as pgn
        ON pgc.relnamespace = pgn.oid
    /* columns */
    /* This is a simplified version of how information_schema.columns was
    ** implemented in postgres 9.5, but modified to support materialized
    ** views.
    */
    LEFT OUTER JOIN pg_attribute AS pga
      ON pga.attrelid = pgc.oid
    LEFT OUTER JOIN (
      SELECT
        nc.nspname         AS table_schema,
        c.relname          AS table_name,
        a.attname          AS column_name,
        a.attnum           AS ordinal_position,
        pg_get_expr(ad.adbin, ad.adrelid) AS column_default,
        CASE WHEN a.attnotnull OR (t.typtype = 'd' AND t.typnotnull) THEN 'NO' ELSE 'YES' END AS is_nullable,
        CASE WHEN t.typtype = 'd' THEN
          CASE WHEN bt.typelem <> 0 AND bt.typlen = -1 THEN 'ARRAY'
               WHEN nbt.nspname = 'pg_catalog' THEN format_type(t.typbasetype, null)
               ELSE 'USER-DEFINED' END
        ELSE
          CASE WHEN t.typelem <> 0 AND t.typlen = -1 THEN 'ARRAY'
               WHEN nt.nspname = 'pg_catalog' THEN format_type(a.atttypid, null)
               ELSE 'USER-DEFINED' END
        END AS data_type,
        coalesce(bt.typname, t.typname) AS data_type_name
      FROM (pg_attribute a LEFT JOIN pg_attrdef ad ON attrelid = adrelid AND attnum = adnum)
        JOIN (pg_class c JOIN pg_namespace nc ON (c.relnamespace = nc.oid)) ON a.attrelid = c.oid
        JOIN (pg_type t JOIN pg_namespace nt ON (t.typnamespace = nt.oid)) ON a.atttypid = t.oid
        LEFT JOIN (pg_type bt JOIN pg_namespace nbt ON (bt.typnamespace = nbt.oid))
          ON (t.typtype = 'd' AND t.typbasetype = bt.oid)
        LEFT JOIN (pg_collation co JOIN pg_namespace nco ON (co.collnamespace = nco.oid))
          ON a.attcollation = co.oid AND (nco.nspname, co.collname) <> ('pg_catalog', 'default')
      WHERE (NOT pg_is_other_temp_schema(nc.oid))
        AND a.attnum > 0 AND NOT a.attisdropped AND c.relkind in ('r', 'v', 'm', 'f', 'p')
        AND (pg_has_role(c.relowner, 'USAGE')
             OR has_column_privilege(c.oid, a.attnum,
                                     'SELECT, INSERT, UPDATE, REFERENCES'))
    ) AS isc
      ON  isc.table_schema = pgn.nspname
      AND isc.table_name   = pgc.relname
      AND isc.column_name  = pga.attname
  
    /* triggers */
    LEFT OUTER JOIN pg_trigger AS pgt
      ON pgt.tgrelid = pgc.oid
    LEFT OUTER JOIN information_schema.triggers AS ist
      ON  ist.event_object_schema = pgn.nspname
      AND ist.event_object_table  = pgc.relname
      AND ist.trigger_name        = pgt.tgname
  
    /* This is a simplified version of how information_schema.views was
    ** implemented in postgres 9.5, but modified to support materialized
    ** views.
    */
    LEFT OUTER JOIN (
      SELECT
        nc.nspname         AS table_schema,
        c.relname          AS table_name,
        CASE WHEN pg_has_role(c.relowner, 'USAGE') THEN pg_get_viewdef(c.oid) ELSE null END AS view_definition,
        CASE WHEN pg_relation_is_updatable(c.oid, false) & 20 = 20 THEN 'YES' ELSE 'NO' END AS is_updatable,
        CASE WHEN pg_relation_is_updatable(c.oid, false) &  8 =  8 THEN 'YES' ELSE 'NO' END AS is_insertable_into,
        CASE WHEN EXISTS (SELECT 1 FROM pg_trigger WHERE tgrelid = c.oid AND tgtype & 81 = 81) THEN 'YES' ELSE 'NO' END AS is_trigger_updatable,
        CASE WHEN EXISTS (SELECT 1 FROM pg_trigger WHERE tgrelid = c.oid AND tgtype & 73 = 73) THEN 'YES' ELSE 'NO' END AS is_trigger_deletable,
        CASE WHEN EXISTS (SELECT 1 FROM pg_trigger WHERE tgrelid = c.oid AND tgtype & 69 = 69) THEN 'YES' ELSE 'NO' END AS is_trigger_insertable_into
      FROM pg_namespace nc, pg_class c
  
      WHERE c.relnamespace = nc.oid
        AND c.relkind in ('v', 'm')
        AND (NOT pg_is_other_temp_schema(nc.oid))
        AND (pg_has_role(c.relowner, 'USAGE')
             OR has_table_privilege(c.oid, 'SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, TRIGGER')
             OR has_any_column_privilege(c.oid, 'SELECT, INSERT, UPDATE, REFERENCES'))
    ) AS isv
      ON  isv.table_schema = pgn.nspname
      AND isv.table_name   = pgc.relname
  
    WHERE
      pgc.relkind IN ('r', 'v', 'f', 'm', 'p')
      ${
        'schemas' in options
          ? getSchemasWhereClause(options.schemas, 'AND')('pgn.nspname')
          : getTablesWhereClause(
              options.tables,
              'AND'
            )({
              name: 'pgc.relname',
              schema: 'pgn.nspname',
            })
      }
    GROUP BY pgc.oid, pgn.nspname, pgc.relname, table_type, isv.*
  ) AS info;
  `;
  },
  primaryKeysInfoSql(options: TableORSchemaArg): string {
    return getKeysSql('PRIMARY KEY', options);
  },
  uniqueKeysSql(options: TableORSchemaArg): string {
    return getKeysSql('UNIQUE', options);
  },
  checkConstraintsSql(options: TableORSchemaArg): string {
    return `
-- test_id = ${'schemas' in options ? 'multi' : 'single'}_check_constraint
SELECT
COALESCE(
  json_agg(
    row_to_json(info)
  ),
  '[]' :: JSON
)
FROM (
SELECT n.nspname::text AS table_schema,
    ct.relname::text AS table_name,
    r.conname::text AS constraint_name,
    ccu.column_name,
    pg_get_constraintdef(r.oid, true) AS "check"
   FROM pg_constraint r
     JOIN pg_class ct ON r.conrelid = ct.oid
     JOIN pg_namespace n ON ct.relnamespace = n.oid
     JOIN information_schema.constraint_column_usage ccu
          ON r.conname = ccu.constraint_name
          AND n.nspname = ccu.constraint_schema
  ${
    'schemas' in options
      ? getSchemasWhereClause(options.schemas)('n.nspname')
      : getTablesWhereClause(options.tables)({
          name: 'ct.relname',
          schema: 'n.nspname',
        })
  }
   AND r.contype = 'c'::"char"
   ) AS info;
`;
  },
  getFKRelations(options: TableORSchemaArg): string {
    return `
-- test_id = ${'schemas' in options ? 'multi' : 'single'}_foreign_key
SELECT
COALESCE(json_agg(row_to_json(info)), '[]'::JSON)
FROM (
SELECT
q.table_schema :: text AS table_schema,
q.table_name :: text AS table_name,
q.constraint_name :: text AS constraint_name,
min(q.ref_table_table_schema :: text) AS ref_table_table_schema,
min(q.ref_table :: text) AS ref_table,
json_object_agg(ac.attname, afc.attname) AS column_mapping,
min(q.confupdtype :: text) AS on_update,
min(q.confdeltype :: text) AS on_delete
FROM
(
  SELECT
    ctn.nspname AS table_schema,
    ct.relname AS table_name,
    r.conrelid AS table_id,
    r.conname AS constraint_name,
    cftn.nspname AS ref_table_table_schema,
    cft.relname AS ref_table,
    r.confrelid AS ref_table_id,
    r.confupdtype,
    r.confdeltype,
    unnest(r.conkey) AS column_id,
    unnest(r.confkey) AS ref_column_id
  FROM
    pg_constraint r
    JOIN pg_class ct ON r.conrelid = ct.oid
    JOIN pg_namespace ctn ON ct.relnamespace = ctn.oid
    JOIN pg_class cft ON r.confrelid = cft.oid
    JOIN pg_namespace cftn ON cft.relnamespace = cftn.oid
  WHERE
    r.contype = 'f' :: "char"
) q
JOIN pg_attribute ac ON q.column_id = ac.attnum
AND q.table_id = ac.attrelid
JOIN pg_attribute afc ON q.ref_column_id = afc.attnum
AND q.ref_table_id = afc.attrelid
${
  'schemas' in options
    ? getSchemasWhereClause(options.schemas)([
        'q.table_schema',
        'q.ref_table_table_schema',
      ])
    : getTablesWhereClause(options.tables)([
        { name: 'q.table_name', schema: 'q.table_schema' },
        { name: 'q.ref_table', schema: 'q.ref_table_table_schema' },
      ])
}
GROUP BY
  q.table_schema,
  q.table_name,
  q.constraint_name
  ) AS info;`;
  },
  getTableColumnsSql,
};
