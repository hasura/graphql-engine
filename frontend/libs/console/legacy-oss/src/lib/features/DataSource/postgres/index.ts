import { Database } from '..';
import {
  getDatabaseConfiguration,
  getTrackableTables,
  getTableColumns,
} from './introspection';

export type PostgresTable = { name: string; schema: string };

export const postgres: Database = {
  introspection: {
    getDatabaseConfiguration,
    getTrackableTables,
    getDatabaseHierarchy: async () => {
      return ['schema', 'name'];
    },
    getTableColumns,
  },
};
