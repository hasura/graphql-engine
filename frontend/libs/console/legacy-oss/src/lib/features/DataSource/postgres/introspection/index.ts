export { getTrackableTables } from './getTrackableTables';
export { getDatabaseConfiguration } from './getDatabaseConfiguration';
export { getTableColumns } from './getTableColumns';
