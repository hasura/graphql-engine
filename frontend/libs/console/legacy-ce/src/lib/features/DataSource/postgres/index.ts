import {
  ChangeDatabaseSchemaProps,
  Database,
  GetDefaultQueryRootProps,
  GetVersionProps,
} from '..';
import { Table } from '../../hasura-metadata-types';
import { runSQL } from '../api';
import { postgresCapabilities } from '../common/capabilities';
import {
  defaultDatabaseProps,
  defaultIntrospectionProps,
} from '../common/defaultDatabaseProps';
import {
  getDatabaseConfiguration,
  getDatabaseSchemas,
  getFKRelationships,
  getSupportedOperators,
  getTrackableFunctions,
  getTableColumns,
  getTablesListAsTree,
  getTrackableTables,
  getIsTableView,
} from './introspection';
import { getTableRows } from './query';
import { consoleDataTypeToSQLTypeMap, consoleScalars } from './utils';

export type PostgresTable = { name: string; schema: string };

const getDropSchemaSql = (schemaName: string) =>
  `drop schema "${schemaName}" cascade;`;

const getCreateSchemaSql = (schemaName: string) =>
  `create schema "${schemaName}";`;

export const postgres: Database = {
  ...defaultDatabaseProps,
  introspection: {
    ...defaultIntrospectionProps,
    getTrackableFunctions,
    getVersion: async ({ dataSourceName, httpClient }: GetVersionProps) => {
      const result = await runSQL({
        source: {
          name: dataSourceName,
          kind: 'postgres',
        },
        sql: `SELECT VERSION()`,
        httpClient,
      });
      return result.result?.[1][0] ?? '';
    },
    getDriverInfo: async () => ({
      name: 'postgres',
      displayName: 'Postgres',
      release: 'GA',
      native: true,
    }),
    getDatabaseConfiguration,
    getDriverCapabilities: async () => Promise.resolve(postgresCapabilities),
    getTrackableTables,
    getDatabaseHierarchy: async () => {
      return ['schema', 'name'];
    },
    getTableColumns,
    getFKRelationships,
    getTablesListAsTree,
    getSupportedOperators,
    getDatabaseSchemas,
    getIsTableView,
    getSupportedDataTypes: async () => consoleDataTypeToSQLTypeMap,
    getSupportedScalars: async () => consoleScalars,
  },
  query: {
    getTableRows,
  },
  modify: {
    defaultQueryRoot: async ({ table }: GetDefaultQueryRootProps) => {
      const { name, schema } = table as PostgresTable;

      return schema === 'public' ? name : `${schema}_${name}`;
    },
    createDatabaseSchema: async ({
      dataSourceName,
      schemaName,
      httpClient,
    }: ChangeDatabaseSchemaProps) => {
      const response = await runSQL({
        source: { name: dataSourceName, kind: 'postgres' },
        sql: getCreateSchemaSql(schemaName),
        httpClient,
      });
      return response;
    },
    deleteDatabaseSchema: async ({
      dataSourceName,
      schemaName,
      httpClient,
    }: ChangeDatabaseSchemaProps) => {
      const response = await runSQL({
        source: { name: dataSourceName, kind: 'postgres' },
        sql: getDropSchemaSql(schemaName),
        httpClient,
      });
      return response;
    },
  },
  config: {
    getDefaultQueryRoot: async (table: Table) => {
      const { name, schema } = table as PostgresTable;
      return schema === 'public' ? name : `${schema}_${name}`;
    },
    getSupportedQueryTypes: async () => {
      return ['select', 'insert', 'update', 'delete'];
    },
  },
};
