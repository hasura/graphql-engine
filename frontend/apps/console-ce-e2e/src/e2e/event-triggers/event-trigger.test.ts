import { HasuraMetadataV3 } from '@hasura/console-legacy-ce';
import { readMetadata } from '../actions/withTransform/utils/services/readMetadata';
import { postgres } from '../data/manage-database/postgres.spec';

describe('Create event trigger with shortest possible path', () => {
  before(() => {
    // create a table first
    postgres.helpers.createTable('user_table');

    // track the table
    cy.visit('/data/default/schema/public', {
      timeout: 10000,
    });
    cy.get('[data-test=add-track-table-user_table]', {
      timeout: 10000,
    }).click();
  });
  after(() => {
    // delete the table
    cy.log('**--- Delete the table');
    postgres.helpers.deleteTable('user_table');
  });

  it('When the users create, modify and delete an Event trigger, everything should work', () => {
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**--- Step 1: Create an ET with shortest path**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');

    cy.visit('/events/data/manage', {
      timeout: 10000,
      onBeforeLoad(win) {
        cy.stub(win, 'prompt').returns('event_trigger_test');
      },
    });

    // --------------------
    cy.log('**--- Click on the Create button of the ET panel**');
    cy.get('[data-testid=data-create-trigger]').click();

    // trigger name
    cy.log('**--- Type the event trigger name**');
    cy.findByPlaceholderText('trigger_name').type('event_trigger_test');

    // select source
    cy.log('**--- Select the DB');
    cy.get('[name="source"]').select('default');

    // select schema and table
    cy.log('**--- Select the schema and table');
    cy.get('[name="schema"]').select('public');
    cy.get('[name="tableName"]').select('user_table');

    // select the trigger operation
    cy.log('**--- Select the ET operation');
    cy.get('[name=insert]').click();

    // add webhook url
    cy.log('**--- Add webhook url');
    cy.get('[name=handler]').type('http://httpbin.org/post');

    // click on create button to save ET
    cy.log('**--- Click on Create Event Trigger');
    cy.findByRole('button', { name: 'Create Event Trigger' }).click();

    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**--- Step 2: Modify an ET to longest path and save it**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');

    // modfy the trigger operation
    cy.log(
      '**--- Click on Edit trigger operation and modfiy the trigger operation'
    );
    cy.findAllByRole('button', { name: 'Edit' }).eq(1).click();
    cy.get('[name=update]').click();
    cy.get('[name=column-id]').click();
    cy.findByRole('button', { name: 'Save' }).click();

    // modify the retry config
    cy.log('**--- Click on Edit retry config and modify the config');
    cy.findAllByRole('button', { name: 'Edit' }).eq(1).click();
    cy.get('[name=num_retries]').clear().type('10');
    cy.get('[name=interval_sec]').clear().type('5');
    cy.get('[name=timeout_sec]').clear().type('70');
    cy.findByRole('button', { name: 'Save' }).click();

    // add headers
    cy.log('**--- Click on Edit retry config and add header');
    cy.findAllByRole('button', { name: 'Edit' }).eq(2).click();
    cy.findByPlaceholderText('key').type('x-hasura-user-id');
    cy.findByPlaceholderText('value').type('1234');
    cy.findByRole('button', { name: 'Save' }).click();

    // add Sample context
    cy.log('**--- Click on show sample context and fill the form');
    cy.findByText('Show Sample Context').click();
    cy.findByPlaceholderText('Key...').type('env-var');
    cy.findByPlaceholderText('Value...').type('env-var-value');

    // add Request Options Transform
    cy.log('**--- Click on Add Request Options Transform and fill the form');
    cy.findByText('Add Request Options Transform').click();
    cy.get('[name=GET]').click();
    cy.get('[name=request_url]').type('/transformUrl');
    cy.findAllByPlaceholderText('Key...').eq(2).type('x-hasura-user-id');
    cy.findAllByPlaceholderText('Value...').eq(2).type('my-user-id');

    // add Payload Transform
    cy.log('**--- Click on Add Payload Transform and fill the form');
    cy.findByText('Add Payload Transform').click();

    // save the ET
    cy.log('**--- Click on Save Event trigger to modify the ET');
    cy.findByRole('button', { name: 'Save Event Trigger' }).click();

    readMetadata().then((md: { body: HasuraMetadataV3 }) => {
      cy.wrap(
        (md.body.sources || [])
          .find(source => source.name === 'default')
          .tables.find(table => table?.table?.name === 'user_table')
          ?.event_triggers?.[0]
      ).toMatchSnapshot({ name: 'Modify the shotest path to longest' });
    });

    // delete ET
    cy.intercept('POST', 'http://localhost:8080/v1/metadata', req => {
      if (JSON.stringify(req.body).includes('delete_event_trigger')) {
        req.alias = 'deleteTrigger';
      }
      req.continue();
    });

    cy.intercept('POST', 'http://localhost:9693/apis/migrate', req => {
      if (JSON.stringify(req.body).includes('delete_event_trigger')) {
        req.alias = 'deleteTrigger';
      }
    });
    cy.findByRole('button', { name: 'Delete Event Trigger' }).click();
    cy.wait('@deleteTrigger');
  });
});

describe('Create event trigger with logest possible path', () => {
  before(() => {
    // create a table first
    postgres.helpers.createTable('user_table');

    // track the table
    cy.visit('/data/default/schema/public', {
      timeout: 10000,
    });
    cy.get('[data-test=add-track-table-user_table]', {
      timeout: 10000,
    }).click();
  });
  after(() => {
    // delete the table
    cy.log('**--- Delete the table');
    postgres.helpers.deleteTable('user_table');
  });
  xit('When the users create, modify and delete an Event trigger, everything should work', () => {
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**--- Step 1: Create an ET with longest path**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');

    cy.visit('/events/data/manage', {
      timeout: 10000,
      onBeforeLoad(win) {
        cy.stub(win, 'prompt').returns('event_trigger_test');
      },
    });

    // --------------------
    cy.log('**--- Click on the Create button of the ET panel**');
    cy.get('[data-testid=data-create-trigger]').click();

    // trigger name
    cy.log('**--- Type the event trigger name**');
    cy.findByPlaceholderText('trigger_name').type('event_trigger_test');

    // select source
    cy.log('**--- Select the DB');
    cy.get('[name="source"]').select('default');

    // select schema and table
    cy.log('**--- Select the schema and table');
    cy.get('[name="schema"]').select('public');
    cy.get('[name="tableName"]').select('user_table');

    // select the trigger operation
    cy.log('**--- Select the ET operation');
    cy.get('[name=insert]').click();

    // add webhook url
    cy.log('**--- Add webhook url');
    cy.get('[name=handler]').type('http://httpbin.org/post');

    // toggle the adv setting, add retry config and headers
    cy.log('**--- Add retry config');
    cy.findByText('Advanced Settings').click();
    cy.get('[name=num_retries]').clear().type('10');
    cy.get('[name=interval_sec]').clear().type('5');
    cy.get('[name=timeout_sec]').clear().type('70');

    // add headers
    cy.log('**--- Add header');
    cy.findByPlaceholderText('key').type('x-hasura-user-id');
    cy.findByPlaceholderText('value').type('1234');

    // add Sample context
    cy.log('**--- Click on show sample context and fill the form');
    cy.findByText('Show Sample Context').click();
    cy.findByPlaceholderText('Key...').type('env-var');
    cy.findByPlaceholderText('Value...').type('env-var-value');

    // add Request Options Transform
    cy.log('**--- Click on Add Request Options Transform and fill the form');
    cy.findByText('Add Request Options Transform').click();
    cy.get('[name=GET]').click();
    cy.get('[name=request_url]').type('/transformUrl');
    cy.findAllByPlaceholderText('Key...').eq(2).type('x-hasura-user-id');
    cy.findAllByPlaceholderText('Value...').eq(2).type('my-user-id');

    // add Payload Transform
    cy.log('**--- Click on Add Payload Transform and fill the form');
    cy.findByText('Add Payload Transform').click();

    // click on create button to save ET
    cy.log('**--- Click on Create Event Trigger');
    cy.get('[data-test=trigger-create]').click();

    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**--- Step 2: Modify an ET to shortest path and save it**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');
    cy.log('**------------------------------**');

    // modfy the trigger operation
    cy.log(
      '**--- Click on Edit trigger operation and modfiy the trigger operation'
    );
    cy.findAllByRole('button', { name: 'Edit' }).eq(1).click();
    cy.get('[name=update]').click();
    cy.get('[name=column-id]').click();
    cy.findByRole('button', { name: 'Save' }).click();

    // remove Request Options Transform
    cy.log('**--- Click on Remove Request Options Transform');
    cy.findByText('Remove Request Options Transform').click();

    // remove Payload Transform
    cy.log('**--- Click on Remove Payload Transform');
    cy.findByText('Remove Payload Transform').click();

    // save the ET
    cy.log('**--- Click on Save Event trigger to modify the ET');
    cy.findByRole('button', { name: 'Save Event Trigger' }).click();

    readMetadata().then((md: { body: HasuraMetadataV3 }) => {
      cy.wrap(
        (md.body.sources || [])
          .find(source => source.name === 'default')
          .tables.find(table => table?.table?.name === 'user_table')
          ?.event_triggers?.[0]
      ).toMatchSnapshot({ name: 'Modify the longest path to shortest path' });
    });

    // delete ET
    cy.intercept('POST', 'http://localhost:8080/v1/metadata', req => {
      if (JSON.stringify(req.body).includes('delete_event_trigger')) {
        req.alias = 'deleteTrigger';
      }
      req.continue();
    });

    cy.intercept('POST', 'http://localhost:9693/apis/migrate', req => {
      if (JSON.stringify(req.body).includes('delete_event_trigger')) {
        req.alias = 'deleteTrigger';
      }
    });
    cy.findByRole('button', { name: 'Delete Event Trigger' }).click();
    cy.wait('@deleteTrigger');
  });
});
