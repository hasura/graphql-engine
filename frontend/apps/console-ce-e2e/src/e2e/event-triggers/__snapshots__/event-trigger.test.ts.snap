exports[`Create event trigger with shortest possible path > When the users create, modify and delete an Event trigger, everything should work #0`] =
{
  "definition": {
    "enable_manual": false,
    "insert": {
      "columns": "*"
    },
    "update": {
      "columns": [
        "id"
      ]
    }
  },
  "headers": [
    {
      "name": "x-hasura-user-id",
      "value": "1234"
    }
  ],
  "name": "event_trigger_test",
  "request_transform": {
    "body": {
      "action": "transform",
      "template": "{\n  \"table\": {\n    \"name\": {{$body.table.name}},\n    \"schema\": {{$body.table.schema}}\n  }\n}"
    },
    "method": "GET",
    "query_params": {
      "x-hasura-user-id": "my-user-id"
    },
    "request_headers": {
      "add_headers": {},
      "remove_headers": [
        "content-type"
      ]
    },
    "template_engine": "Kriti",
    "url": "{{$base_url}}/transformUrl",
    "version": 2
  },
  "retry_conf": {
    "interval_sec": 5,
    "num_retries": 10,
    "timeout_sec": 70
  },
  "webhook": "http://httpbin.org/post"
};